nightmare_firmware/nightmare_firmware.ino

```diff
@@ -39,17 +42,109 @@ void setup(){
 }
 
 void loop(){
-	/* esp32 bound packet (ROS -> esp32):
-		>HEADER: (ten 0x55 bytes by default)
-		>SERVO ANGLES: an array of 27 angles, one byte each going from 0 to 240 meaning -120 to 120 
+	read_packet();
+	write_packet();
+}
+
+/**
+ * @brief function to send ROS bound packet
+ * @note ROS bound packet (esp32 -> ROS):
+ * 
+		## BASIC DATA PACKET ##
+
+			{HEADER + BATTERY VOLTAGE + BATTERY CURRENT + JOYSTICK POS + JOYSTICK STATE + CHECKSUM}
+
+			total bytes: 10 (if header not changed) + 1 + 1 + 4 + 1 + 1 = 18 bytes
+			checksum bytes: 1 + 1 + 4 + 1 + 1 = 8 bytes
+			
+		## FULL DATA PACKET ##
+
+			{HEADER + SERVO ANGLES + BATTERY VOLTAGE + BATTERY CURRENT + JOYSTICK POS + JOYSTICK STATE + CHECKSUM}
+
+			total bytes: 10 (if header not changed) + 54 + 1 + 1 + 4 + 1 + 1 = 72 bytes
+			checksum bytes: 54 + 1 + 1 + 4 + 1 + 1 = 62 bytes
+			
+		>HEADER -> (ten 0x55 bytes by default)
+		>ID -> 0 for basic data, 1 for full data
+		>SERVO ANGLES ->  an array of 27 angles, two bytes each going from 0 to 1500 meaning -180 to 180 
+					  	the last angle (tilt of the head) is in range -90..90 because it is a standard servo
+					  	and not a serial one but still sent as 2 bytes
+		>BATTERY VOLTAGE -> one byte 0..255 meaning 0..9V
+		>BATTERY CURRENT -> one byte 0..255 meaning 0..12A
+		>JOYSTICK POS -> 4 bytes describing the two joystick positions in the following order: LX LY RX RY
+		>JOYSTICK STATE -> 1 byte sent by the nightcontrol app through 433Mhz
+		>CHECKSUM -> the first 8 bits of the sum of all the packet's bytes
+ */
+void write_packet(){
+	/* check if all servos are attached */
+	bool all_active = true;
+	for(auto& servo : servos){
+		if(servo.active == false){
+			all_active = false;
+		}
+	}
+	if(!hservo.attached()){
+		all_active = false;
+	}
+
+	int header_size = sizeof(HEADER)/sizeof(HEADER[0]);
+	if(all_active){
+		// send basic data
+		uint8_t buf[18] = {};
+		for(int i=0; i<header_size; i++){
+			buf[i] = HEADER[i];
+		}
+		buf[10] = 0; //TODO battery voltage
+		buf[11] = 0; //TODO battery current
+		buf[12] = 0; //TODO joystick LX
+		buf[13] = 0; //TODO joystick LY
+		buf[14] = 0; //TODO joystick RX
+		buf[15] = 0; //TODO joystick RY
+		buf[16] = 0; //TODO joystick state
+		
+		// calculate checksum
+		int checksum = 0;
+		for(auto& buf_byte : buf){
+			checksum += buf_byte;
+		}
+		
+		buf[17] = (uint8_t)checksum;
```

Comment:
Shouldn't this be `buf[17] = ~(uint8_t)checksum;` ?

Reply:
no the ~ comes from the servo protocol that uses the inverted checksum here I'm simply using the sum for simplicity